athena/etc/desync/desync.8
.\" $Id: desync.8,v 1.2 2000-09-23 19:54:35 ghudson Exp $
.\"
.\" Copyright 1995, 1996, 1997 by the Massachusetts Institute of Technology.
.\"
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies and that both that copyright
.\" notice and this permission notice appear in supporting
.\" documentation, and that the name of M.I.T. not be used in
.\" advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\" M.I.T. makes no representations about the suitability of
.\" this software for any purpose.  It is provided "as is"
.\" without express or implied warranty.
.\"
.TH DESYNC 1 "5 March 1997"
.SH NAME
desync \- desynchronize timed jobs on networks
.SH SYNOPSIS
.B desync
[
.B \-t
timefile ] [ range ]
.SH DESCRIPTION
.I desync
is a tool which sleeps a random (hostname seeded) period of time (up
to an hour, by default) in order to skew the network load from several
machines running
.I cron
jobs that would otherwise be synchronized.

An invocation would appear something like

.nf
     42 * * * *          desync; /usr/lib/sendmail -q
.fi

in a crontab file.
.SH OPTIONS
.TP 8
.B range
This optional range specifies, in seconds, the maximum value for the
randomly selected sleep time.  The default value is 3600.
.TP 8
.B \-t timefile
This option changes the behavior of
.I desync
to make it useful for desynchronizing jobs when sleeping is not
appropriate.  If
.I timefile
does not exist,
.I desync
selects a random sleep time using the normal method, adds it to the
current time value, writes the result into
.IR timefile ,
and exits with status 1.  If
.I timefile
does exist and the current time is less than the time value listed in
.IR timefile ,
then
.I desync
takes no action and exits with status 1.  If
.I timefile
exists and the current time is equal to or greater than the time value
listed in
.IR timefile ,
then
.I desync
unlinks
.I timefile
and exits with status 0.  This feature can be used in shell code of
the following form:

.nf
	if desync -t /tmp/mytimefile 3600; then
		echo "Starting the job."
	else
		echo "Putting it off until later."
	fi
.fi

.SH DIAGNOSTICS
If
.I desync
fails to acquire the machine's IP address from
.IR rc.conf ,
it returns an exit status of 2 and does not sleep.
.SH FILES
.TP 25
.B /etc/athena/rc.conf
Athena startup configuration file
.SH SEE ALSO
cron(1)
.SH AUTHOR
Craig Fields, MIT Information Systems
.br
Greg Hudson, MIT Information Systems
.br
Copyright (c) 1995, 1996, 1997, Massachusetts Institute of Technology


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


athena/etc/desync/desync.c
/* Copyright 1996, 1997 by the Massachusetts Institute of Technology.
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

static const char rcsid[] = "$Id: desync.c,v 1.7.2.1 2000-09-23 19:32:31 ghudson Exp $";

/*
 * desync - desynchronize cron jobs on networks
 *
 * This program is a tool which sleeps an ip-address dependent period
 * of time in order to skew over the course of a set time cron jobs
 * that would otherwise be synchronized. It should reside on local disk
 * so as not to cause a fileserver load at its own invocation.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <ctype.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <time.h>

extern int optind;
extern char *optarg;

static char *progname;

static unsigned long get_hostname_hash(void);

int main(int argc, char **argv)
{
  const char *timefile = NULL;
  int range, interval, c;
  unsigned long tval, seed;
  FILE *fp;

  /* Save the program name. */
  progname = argv[0];

  /* Parse command-line flags. */
  while ((c = getopt(argc, argv, "t:")) != -1)
    {
      switch (c) {
      case 't':
	timefile = optarg;
	break;
      default:
	fprintf(stderr, "Usage: %s [-t timefile] [interval]\n", progname);
	return 2;
      }
    }

  /* Get the time interval from the remaining argument, if there is one. */
  argc -= optind;
  argv += optind;
  range = (argc == 1) ? atoi(argv[0]) : 3600;

  seed = get_hostname_hash();
  if (seed == 0)
    return 2;

  srand(seed);
  interval = rand() % range;

  if (timefile)
    {
      fp = fopen(timefile, "r");
      if (fp)
	{
	  if (fscanf(fp, "%lu", &tval) != 1)
	    {
	      fprintf(stderr, "%s: Invalid time file %s\n", progname,
		      timefile);
	      return 2;
	    }
	  fclose(fp);
	  if (time(NULL) >= tval)
	    {
	      unlink(timefile);
	      return 0;
	    }
	  else
	    return 1;
	}
      else
	{
	  if (interval == 0)
	    return 0;
	  fp = fopen(timefile, "w");
	  if (fp == NULL)
	    {
	      fprintf(stderr, "%s: Couldn't open %s for writing: %s\n",
		      progname, timefile, strerror(errno));
	      return 2;
	    }
	  fprintf(fp, "%lu\n", (unsigned long)(time(NULL) + interval));
	  fclose(fp);
	  return 1;
	}
    }
  else
    sleep(interval);

  return 0;
}

static unsigned long get_hostname_hash(void)
{
  char buf[128], *p;
  unsigned long g, hashval = 0;

  if (gethostname(buf, sizeof(buf)) < 0)
    {
      fprintf(stderr, "%s: Unable to obtain hostname: %s\n",
	      progname, strerror(errno));
      return 0;
    }
  buf[sizeof(buf) - 1] = '\0';

  for (p = buf; *p; p++)
    {
      hashval = (hashval << 4) + *p;
      g = hashval & 0xf0000000;
      if (g != 0) {
	hashval ^= g >> 24;
	hashval ^= g;
      }
    }

  if (hashval == 0)
    hashval = 1;

  return hashval;
}
